controllers/cluster_controller.go

```diff
@@ -18,34 +18,230 @@ package controllers
 )
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
 	client.Client
 	Log logr.Logger
+
+	controller       controller.Controller
+	recorder         record.EventRecorder
+	externalWatchers sync.Map
 }
 
-// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;update;patch
+func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		Build(r)
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("cluster-controller")
+	return err
+}
 
-func (r *ClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
+func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
 	_ = r.Log.WithValues("cluster", req.NamespacedName)
 
-	// your logic here
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	err := r.Get(ctx, req.NamespacedName, cluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Store Cluster early state to allow patching.
+	patchCluster := client.MergeFrom(cluster.DeepCopy())
+
+	// Always issue a Patch for the Cluster object and its status after each reconciliation.
+	defer func() {
+		gvk := cluster.GroupVersionKind()
+		if err := r.Client.Patch(ctx, cluster, patchCluster); err != nil {
+			klog.Errorf("Error Patching Cluster %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+			if reterr == nil {
+				reterr = err
+			}
+			return
+		}
+		// TODO(vincepri): This is a hack because after a Patch, the object loses TypeMeta information.
+		// Remove when https://github.com/kubernetes-sigs/controller-runtime/issues/526 is fixed.
+		cluster.SetGroupVersionKind(gvk)
+		if err := r.Client.Status().Patch(ctx, cluster, patchCluster); err != nil {
+			klog.Errorf("Error Patching Cluster status %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+			if reterr == nil {
+				reterr = err
+			}
+		}
+	}()
+
+	// If object hasn't been deleted and doesn't have a finalizer, add one.
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !util.Contains(cluster.Finalizers, clusterv1.ClusterFinalizer) {
+			cluster.Finalizers = append(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer)
+		}
+	}
+
+	if err := r.reconcile(ctx, cluster); err != nil {
+		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
+			klog.Infof("Reconciliation for Cluster %q in namespace %q asked to requeue: %v", cluster.Name, cluster.Namespace, err)
```

Comment:
:-(

Reply:
(on the use of klog directly over logr)

Reply:
Yeah! I'll do it in a later iteration, I want to minimize the amount of changes to get to Kubebuilder V2, and optimize later